proposals/p2173.md

```diff
@@ -0,0 +1,489 @@
+  let Element:! Type;
+  let Slice:! Container where .Element = Element;
+}
+// `T.Slice.Element` rewritten to `T.Element` because type of `T.Slice` says `.Element = Element`.
+// `T.Element` rewritten to `i32` because type of `T` says `.Element = i32`.
+fn A[T:! Container where .Element = i32](x: T.Slice.Element) {}
+```
+
+Rewrites aren't performed on the left-hand side of such an `=`, so
+`where .A = .B and .A = C` is not rewritten to `where .A = .B and .B = C`.
+Instead, such a `where` clause is invalid unless each rule for `.A` specifies
+the same rewrite.
+
+When an `=` appears within an interface, it is possible to form a cycle where a
+type becomes more complex with each iteration. In order to avoid this problem,
+if, when substituting into the right-hand side of such an `=`, the same rewrite
+is performed again, the program is invalid.
+
+```
+interface Container {
+  let Element:! Type;
+  // (1), bad constraint
+  let Slice:! Container where .Element = .Slice.Element;
+}
+// Error, `T.Slice.Element` rewritten to
+// `T.Slice.Slice.Element` by (1), then to
+// `T.Slice.Slice.Slice.Element` by (1)
+fn A[T:! Container](x: T.Slice.Element) {}
+```
+
+It is valid for the same rewrite rule to be used multiple times within a type
+expression, so long as it is not used during substitution into its own
+right-hand side:
+
+```
+interface BInterface {
+  let C:! Type;
+}
+interface AInterface {
+  let D:! AInterface;
+  let B:! BInterface where .C = Self.D;
+}
+// `A.B.C` is rewritten to `Self.D`, because the type of `A.B` says `.C = Self.D`.
+//     `Self`=`A` is substituted into this giving `A.D`.
+//     During this substitution, the same rewrite rule is not used again,
+//     so this is OK.
+// Then `A.D.B.C` is rewritten to `Self.D`, because the type of `A.D.B` says
+// `.C = Self.D`.
+//     `Self`=`A.D` is substituted into this giving `A.D.D`.
+//     During this substitution, the same rewrite rule is not used again,
+//     so this is also OK, even though the same rule was used twice
+//     in forming the overall type.
+fn F(A:! AInterface, d: A.B.C.B.C) {}
+```
+
+Rewrite constraints thereby give us a deterministic, terminating type
+canonicalization mechanism for associated constants: in `A.B`, if the type of
+`A` specifies that `.B == C`, then `A.B` is replaced by `C`. Equality of types
+constrained in this way is transitive.
+
+Note that given `T:! C where .R = i32` can result in a type `T.R` whose behavior
+is different from the behavior of `T.R` given `T:! C`. For example, member
+lookup into `T.R` can find different results and operations can therefore have
+different behavior. However, this does not violate coherence because `C` and
+`C where .R = i32` don't differ by merely having more type information; rather,
+they are different types that have an isomorphic set of values, somewhat like
+`i32` and `u32`. An `=` constraint is not merely learning a new fact about a
+type, it is requesting different behavior.
+
```

Comment:
More notes:
- _Uses_ are error, not sets of constraints (not ideal, but may be necessary to make this efficient for the compiler)
- Interaction between `=` and `is`
- Explicitly an experiment to see when `==` is needed. Recommend using `=` over `==` where possible.
- An `=` constraint identified with the source code that defines it, for purposes of defining whether a constraint is reused. Different parameters for interface does not mean different `where` clause.